### PR TITLE
os/bluestore: enlarege aligned_size avoid too many vector(> IOV_MAX)

### PR DESCRIPTION
**Pull request by majianpeng**
Fixes: http://tracker.ceph.com/issues/21932
Signed-off-by: Jianpeng Ma <jianpeng.ma@intel.com>

### CODE REVIEW

File: src/os/bluestore/KernelDevice.cc [hunk 1]
```diff
@@ -592,9 +592,15 @@ int KernelDevice::write(
...
   assert(is_valid_io(off, len));
 
-  if ((!buffered || bl.get_num_buffers() >= IOV_MAX) &&
-      bl.rebuild_aligned_size_and_memory(block_size, block_size)) {
-    dout(20) << __func__ << " rebuilding buffer to be aligned" << dendl;
+  if ((!buffered || bl.get_num_buffers() >= IOV_MAX)) {
+    uint64_t aligned_size = block_size;
+    // Avoid after rebuild_aligned_size_and_memory, get_num_buffers() > IOV_MAX
+    if (bl.get_num_buffers() >= IOV_MAX && (len > (IOV_MAX * block_size))) {
+      aligned_size = ROUND_UP_TO(len / IOV_MAX, block_size);
```
Comment (tchaikov) [inline src/os/bluestore/KernelDevice.cc]: what if len is not the multiple of `(IOV_MAX * block_size)`? i think it should be
`ROUND_UP_TO(ROUND_UP_TO(len, IOV_MAX) / IOV_MAX, block_size)`

Comment (xiexingguo) [reply src/os/bluestore/KernelDevice.cc]: +1

File: src/os/bluestore/KernelDevice.cc [hunk 2]
```diff
@@ -592,9 +592,15 @@ int KernelDevice::write(
...
+    }
+    if (bl.rebuild_aligned_size_and_memory(aligned_size, block_size)) {
+      dout(20) << __func__ << " rebuilding buffer to be aligned" << dendl;
```
Comment (tchaikov) [inline src/os/bluestore/KernelDevice.cc]: could you print out the `aligned_size` here also? i think it could help with the performance related diagnosis.

**Comment by majianpeng** [comment]
update.

File: src/os/bluestore/KernelDevice.cc [hunk 3]
```diff
@@ -592,9 +592,17 @@ int KernelDevice::write(
 	   << dendl;
   assert(is_valid_io(off, len));
 
-  if ((!buffered || bl.get_num_buffers() >= IOV_MAX) &&
-      bl.rebuild_aligned_size_and_memory(block_size, block_size)) {
-    dout(20) << __func__ << " rebuilding buffer to be aligned" << dendl;
+  if ((!buffered || bl.get_num_buffers() >= IOV_MAX)) {
+    uint64_t aligned_size = block_size;
+    // Avoid after rebuild_aligned_size_and_memory, get_num_buffers() > IOV_MAX
+    if (bl.get_num_buffers() >= IOV_MAX && (len > (IOV_MAX * block_size))) {
```
Comment (tchaikov) [inline src/os/bluestore/KernelDevice.cc]: to be more consistent, could you please remove the parenthesis around `len > (IOV_MAX * block_size)` ?

File: src/os/bluestore/KernelDevice.cc [hunk 4]
```diff
@@ -592,9 +592,17 @@ int KernelDevice::write(
...
+      aligned_size = ROUND_UP_TO(ROUND_UP_TO(len, IOV_MAX) / IOV_MAX, block_size);
+      dout(5) << __func__ << " because too many buffers, using  " << aligned_size
```
Comment (tchaikov) [inline src/os/bluestore/KernelDevice.cc]: i am not quite sure about `dout(5)`. is a large buffer size an unusual thing and we should watch out? actually what i suggested in last comment is to print the `aligned_size` at the "rebuilding buffer to be aligned" line with `dout(20)`.

**Comment by majianpeng** [comment]
update.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2017-11-10_06:12:07-rados-wip-kefu-testing-2017-11-10-1154-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2017-11-12_17:42:28-rados-wip-kefu-testing-2017-11-10-1154-distro-basic-mira/

**Pull request merged**